core/system/src/page_allocator.c

```diff
@@ -0,0 +1,287 @@
+    /* This is the page heap start address */
+    g_page_heap_start = (void *) start;
+
+    /* How many pages can we fit in here? */
+    if ((start + g_page_size) > (uint32_t) heap_end) {
+        g_page_count_total = 0;
+    }
+    else {
+        g_page_count_total = ((uint32_t) heap_end - start) / g_page_size;
+    }
+    /* Clamp page count to table size */
+    if (g_page_count_total > UVISOR_PAGE_TABLE_MAX_COUNT) {
+        g_page_count_total = UVISOR_PAGE_TABLE_MAX_COUNT;
+    }
+    g_page_count_free = g_page_count_total;
+    /* Remember the end of the heap */
+    g_page_heap_end = g_page_heap_start + g_page_count_free * g_page_size;
+
+    DPRINTF("uvisor_page_init:\n.page_heap start 0x%08x\n.page_heap end   0x%08x\n.page_heap available %ukB split into %u pages of %ukB\n\n",
+            (unsigned int) g_page_heap_start,
+            (unsigned int) g_page_heap_end,
+            (unsigned int) (g_page_count_free * g_page_size / 1024),
+            (unsigned int) g_page_count_total,
+            (unsigned int) (g_page_size / 1024));
+
+    size_t page = 0;
+    for (; page < g_page_count_total; page++) {
+        g_page_owner_table[page] = UVISOR_PAGE_UNUSED;
+    }
+}
+
+int page_allocator_malloc(UvisorPageTable * const table)
+{
+    uint32_t pages_required = page_table_read((uint32_t) &(table->page_count));
+    uint32_t page_size = page_table_read((uint32_t) &(table->page_size));
+    /* Check if the user even wants any pages */
+    if (pages_required == 0) {
+        DPRINTF("uvisor_page_malloc: FAIL: No pages requested!\n\n");
+        return UVISOR_ERROR_PAGE_INVALID_PAGE_COUNT;
+    }
+    /* Check if we can fulfill the requested page size */
+    if (page_size != g_page_size) {
+        DPRINTF("uvisor_page_malloc: FAIL: Requested page size %uB is not the configured page size %uB!\n\n", page_size, g_page_size);
+        return UVISOR_ERROR_PAGE_INVALID_PAGE_SIZE;
+    }
+    /* Check if we have enough pages available */
+    if (pages_required > g_page_count_free) {
+        DPRINTF("uvisor_page_malloc: FAIL: Cannot serve %u pages with only %u free pages!\n\n", pages_required, g_page_count_free);
+        return UVISOR_ERROR_PAGE_OUT_OF_MEMORY;
+    }
+
+    /* Get the calling box id */
+    const page_owner_t box_id = g_active_box;
+    DPRINTF("uvisor_page_malloc: Requesting %u pages with size %uB for box %u\n", pages_required, page_size, box_id);
+
+    /* Update the free pages count */
+    g_page_count_free -= pages_required;
+    /* point to the first entry in the table */
+    void * * page_table = table->page_origins;
+
+    /* Iterate through the page table and find the empty pages */
+    size_t page = 0;
+    for (; (page < g_page_count_total) && pages_required; page++) {
+        /* If the page is unused, it's entry is UVISOR_PAGE_UNUSED (not NULL!) */
+        if (g_page_owner_table[page] == UVISOR_PAGE_UNUSED) {
+            /* Marry this page to the box id */
+            g_page_owner_table[page] = box_id;
```

Comment:
This assumes being called in an SVC, which makes this mutually exclusive.
But, since this same file is used for unsupported, this does indeed lead to a race condition.